### PR TITLE
fix: can not configure upstream with no nodes

### PR DESCRIPTION
**Pull request by liuxiran**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [X] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Please update this section with detailed description.

During rechecking #1803 , there is a problem occured when I try to configure a upstream with no nodes.

Test steps:

1. Create an upstream with no nodes
2. Configure this upstream with no changes, an Submit the data, the error occured

2021-04-21 15-32-32屏幕截图

This pr tried to change the fe logic to solve this problem.

Doc: 
Delete `k8s_deployment_info` in upstream just refer to https://github.com/apache/apisix/blob/359e0906c5a3f0149c0f6a44aff927d4c4eb8c6b/CHANGELOG.md#220

TODO

- [ ] add test case

**Related issues**

None

**Checklist:**

- [X] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: web/src/components/Upstream/service.ts [hunk 1]
```diff
@@ -68,7 +68,6 @@ export const convertToRequestData = (
     type,
     hash_on,
     key,
-    k8s_deployment_info,
```
Comment (juzhiyuan) [inline web/src/components/Upstream/service.ts:71]: yes, we don't support k8s_deployemnt_info here

**Comment by imjoey** [comment]
@liuxiran

How about we make it in the backend? Please see https://github.com/apache/apisix-dashboard/blob/master/api/internal/core/entity/format.go#L49, we can initialize the `nodes` as an empty slice instead of a nil one. Just change `var nodes []*Node` to `nodes := make([]*Node, 0)` .

Looking forward to your insights.

**Comment by liuxiran** [comment]
> @liuxiran
> 
> How about we make it in the backend? Please see https://github.com/apache/apisix-dashboard/blob/master/api/internal/core/entity/format.go#L49, we can initialize the `nodes` as an empty slice instead of a nil one. Just change `var nodes []*Node` to `nodes := make([]*Node, 0)` .
> 
> Looking forward to your insights.

@imjoey Thanks for your advice, and it really make sense for me.

before this change, when get info about an upstream with no nodes
fe will get
WechatIMG5


actually `nodes:null` causes the problem

after this change, fe will get
WechatIMG4

which is matches expectation.

BTW, I have a question in the process of debugging:
there are some logs I added to watch the changes of `upstream.Nodes` before the advice
WechatIMG7
I found even if I did not do this change, the `upstram.Nodes` will always be `[]`, but what Fe got is different, I haven't thought about it for a long time. 

Also @nic-chen @Jaycean @starsz to take a look, thanks a lot

**Comment by starsz** [comment]
> > @liuxiran
> > How about we make it in the backend? Please see https://github.com/apache/apisix-dashboard/blob/master/api/internal/core/entity/format.go#L49, we can initialize the `nodes` as an empty slice instead of a nil one. Just change `var nodes []*Node` to `nodes := make([]*Node, 0)` .
> > Looking forward to your insights.
> 
> @imjoey Thanks for your advice, and it really make sense for me.
> 
> before this change, when get info about an upstream with no nodes
> fe will get
> WechatIMG5
> 
> actually `nodes:null` causes the problem
> 
> after this change, fe will get
> WechatIMG4
> 
> which is matches expectation.
> 
> BTW, I have a question in the process of debugging:
> there are some logs I added to watch the changes of `upstream.Nodes` before the advice
> WechatIMG7
> I found even if I did not do this change, the `upstram.Nodes` will always be `[]`, but what Fe got is different, I haven't thought about it for a long time.
> 
> Also @nic-chen @Jaycean @starsz to take a look, thanks a lot

Hi, @liuxiran 
In golang, you should distinguish `nil slice` and `empty slice`

```
var x []*entity.Node
```
is a nil slice.

```
y := make([]*entity.Node, 0)
```
is an empty slice. The length and value are 0.
If you use printf or println to print them, the result is both "[]".

In this situation, because the upstream.Nodes' type is interface{}.So it's a little more complex.
You can use 
```
reflect.ValueOf(upstream.Nodes).IsNil()
```
to see the true value.

**Comment by liuxiran** [comment]
> Hi, @liuxiran
> In golang, you should distinguish `nil slice` and `empty slice`
> 
> ```
> var x []*entity.Node
> ```
> 
> is a nil slice.
> 
> ```
> y := make([]*entity.Node, 0)
> ```
> 
> is an empty slice. The length and value are 0.
> If you use printf or println to print them, the result is both "[]".
> 
> In this situation, because the upstream.Nodes' type is interface{}.So it's a little more complex.
> You can use
> 
> ```
> reflect.ValueOf(upstream.Nodes).IsNil()
> ```
> 
> to see the true value.

@starsz Thanks for your detailed explanation, after added `reflect.ValueOf(upstream.Nodes).IsNil()` and debug again, I find out the confusion before:
debug step:
1. create an upstream with no nodes
2. visit uptream page, call `GET /apisix/admin/upstreams`
3. config the upstream ,call `GET /apisix/admin/upstream/:id`

before the advice change, I got the logs:
2021-04-23 14-32-00屏幕截图

after the advice change, I got the logs:
2021-04-23 14-39-50屏幕截图

The conclusion is that: 
the original `upstream.nodes` type got from etcd is `[]interface{}`, and it is an empty slice, 
it will be changed to type nodes , `[]*Node`, in FormatNodes `case []interface{}` when I visit upstream page(step2), 
and  the when I got the upstream info(step3), the data come from catch instead of stroge, that is to say, in this scene, the `upstream.nodes` type is already `[]*Node`.

so the advice will solve the problem. 
Thanks again for all your help, then I will try to add related test cases

**Review by juzhiyuan — APPROVED** [review]

File: api/internal/core/entity/format_test.go [hunk 2]
```diff
@@ -101,3 +101,53 @@ func TestNodesFormat_Map(t *testing.T) {
...
+
+	// nodes format
+	nodes := NodesFormat(route.Upstream.Nodes)
+
+	// json encode for client
+	res, err := json.Marshal(nodes)
+	assert.Nil(t, err)
+	jsonStr := string(res)
+	assert.Contains(t, jsonStr, `[]`)
+}
```
Comment (starsz) [inline api/internal/core/entity/format_test.go:153]: Need to test without the "nodes" field.

**Review by starsz — COMMENTED** [review]
Backend LGTM.

File: api/internal/core/entity/format_test.go (hunk 2 above)
Comment (liuxiran) [reply api/internal/core/entity/format_test.go:153]: done.

**Review by LiteSun — APPROVED** [review]
FE LGTM.

**Review by nic-chen — APPROVED** [review]

**Review by imjoey — APPROVED** [review]
/lgtm

**Comment by imjoey** [comment]
@liuxiran FE ci failed after this pr got merged. Could you please take a look, please refer to https://github.com/apache/apisix-dashboard/runs/2430236911. Thanks.

**Comment by liuxiran** [comment]
> @liuxiran FE ci failed after this pr got merged. Could you please take a look, please refer to https://github.com/apache/apisix-dashboard/runs/2430236911. Thanks.

got it, I will fix it later, thanks

**Pull request merged**